File: greedy/python/406.queue-reconstruction-by-height.py
#
# @lc app=leetcode id=406 lang=python3
#
# [406] Queue Reconstruction by Height
#

# @lc code=start
class Solution:
    def reconstructQueue(self, people: List[List[int]]) -> List[List[int]]:
        # sort by height desc, then k asc
        people.sort(key=lambda x: (-x[0], x[1]))
        print(people)

        target_order = [people[0]]

        for i in range(1, len(people)):
            target_order.insert(people[i][1], people[i])

        return target_order
# @lc code=end

File: greedy/cpp/406.queue-reconstruction-by-height.cpp
/*
 * @lc app=leetcode id=406 lang=cpp
 *
 * [406] Queue Reconstruction by Height
 */

// @lc code=start
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> reconstructQueue(vector<vector<int>>& people) {
        sort(people.begin(), people.end(), [](const vector<int> &a, const vector<int> &b) {
            return a[0] > b[0] || (a[0] == b[0] && a[1] < b[1]);
        });

        vector<vector<int>> sortedArray;

        for (auto &p: people) {
            sortedArray.insert(sortedArray.begin() + p[1], p);
        }

        return sortedArray;

    }
};
// @lc code=end

